### PR TITLE
Implement console#time and console#timeEnd methods

### LINKED ISSUE #9325: Implement console#time and console#timeEnd methods
https://developer.mozilla.org/en-US/docs/Web/API/Console#Timers

An effort to implement this was started in https://github.com/servo/servo/pull/6646, but wasn't completed

### PR DESCRIPTION
**Pull request by shssoichiro**
Fixes #9325

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/9420)

### CODE REVIEW

File: components/script/dom/console.rs [hunk 1]
```diff
@@ -92,6 +97,40 @@ impl ConsoleMethods for Console {
...
+            return;
+        }
+        if timers.len() >= 10000 {
+            // Too many timers on page
+            return;
+        }
+
+        timers.insert(label.clone(), timestamp_in_ms(get_time()));
+        let message = DOMString::from(format!("{}: timer started", label));
+        println!("{}", message);
```
Comment (KiChjang) [inline components/script/dom/console.rs]: Could the previous two lines be combined?

File: components/script/dom/console.rs [hunk 2]
```diff
@@ -92,6 +97,40 @@ impl ConsoleMethods for Console {
...
+    }
+
+    // https://developer.mozilla.org/en-US/docs/Web/API/Console/timeEnd
+    fn TimeEnd(&self, label: DOMString) {
+        let mut timers = self.timers.borrow_mut();
+        if let Some(start) = timers.remove(&label) {
+            let message = DOMString::from(
+                format!("{}: {}ms", label, timestamp_in_ms(get_time()) - start)
+            );
+            println!("{}", message);
```
Comment (KiChjang) [inline components/script/dom/console.rs]: Same here. Can the previous two lines be combined?

File: components/script/dom/console.rs (hunk 1 above)
Comment (KiChjang) [inline components/script/dom/console.rs]: Nevermind, I see why you are doing this.

File: components/script/dom/console.rs (hunk 2 above)
Comment (KiChjang) [inline components/script/dom/console.rs]: Ignore my previous message.

**Comment by KiChjang** [comment]
The previous attempt #6646 implements the actual timer hash map on the document object. Why have you instead moved the hash map to the console struct instead?

**Comment by shssoichiro** [comment]
I found it simplest to implement the HashMap on the console struct--since the methods were implemented in that location as well. That being said, it may be more correct to move the HashMap to the document struct--although never explicitly stated in the docs, the note "Up to 10,000 timers can run simultaneously on a given page." does imply that timers should be local to a document.

**Comment by nox** [comment]
Not in the document struct. Workers have access to Console too.

**Comment by shssoichiro** [comment]
Would it be better to leave the `timers` HashMap in the Console then, or to move it into the Document and the Worker? Or move it to a different place entirely? It'd clearly be simpler implementation-wise to keep it in one place, but I'd like to know if there's a benefit to storing the timers in the document/worker.

**Comment by KiChjang** [comment]
I'm inclined to move them into Window and Worker, since that's what `GlobalRef` references.

**Comment by jdm** [comment]
I don't really understand the motivation for that move, since it makes the code more complicated and spread out.

**Comment by nox** [comment]
@jdm How would it work otherwise from a worker?

**Comment by jdm** [comment]
@nox Doesn't the code work as written, by storing all timer-related data in the Console object and not interacting with the global at all?

**Comment by nox** [comment]
Oh sorry, yes, if they live on the Console struct itself that's just fine.

**Comment by KiChjang** [comment]
Okay, looks like we can just accept the code as-is then. Thanks for working on this!

@bors-servo r+

**Comment by KiChjang** [comment]
@bors-servo retry
* #9460

**Pull request merged**